orttraining/orttraining/lazy_tensor/fusion.cc

```diff
@@ -0,0 +1,405 @@
+      getSubgraph(group).registerOutput(mergedNode->output(i));
+      auto new_outer_output = group->addOutput();
+      // Copy metadata from old outer output to new outer output.
+      new_outer_output->copyMetadata(n->output(i));
+      aliasDb_->replaceWithNewValue(n->output(i), new_outer_output);
+    }
+    // Now group is a single-op subgraph containing the clone of n.
+    AT_ASSERT(n->outputs().size() == group->outputs().size());
+    n->replaceAllUsesWith(group);
+    n->destroy();
+    return group;
+  }
+
+  at::optional<torch::jit::Node*> tryFuse(torch::jit::Node* consumer, torch::jit::Value* producer) {
+    // this handles cases where producer can be moved _into_ the fusion group of
+    // consumer.
+    // TODO: extend to fusion of consumer into _producer's_ fusion blob
+    // if the consumer allInputsAreThisProducer(consumer,producer)
+    // we can move the consumer up into the producer.
+    // but this requires better handling of merging fusion groups so it is not
+    // done now
+    bool shouldFuse = isFusable(producer->node()) &&
+                      // Rearrange nodes such that all uses of producer are after the
+                      // consumer. Fusion will rewrite those later uses to use the version of
+                      // producer generated by the fused blob. In this case, producer becomes
+                      // an output of the fusion group.
+                      aliasDb_->moveBeforeTopologicallyValid(producer->node(), consumer);
+
+    if (!shouldFuse) {
+      return at::nullopt;
+    }
+
+    if ((consumer->inputs().size() + consumer->outputs().size() +
+         producer->node()->inputs().size() +
+         producer->node()->outputs().size()) > subgraph_arg_limit_) {
+      return at::nullopt;
+    }
+
+    auto group = consumer;
+    if (consumer->kind() != kind_) {
+      group = createSingletonFusionGroup(consumer);
+    }
+
+    if (producer->node()->kind() == kind_) {
+      mergeFusionGroups(group, producer->node());
+      return group;
+    }
+    // AT_ASSERT(producer->node()->outputs().size() == 1);
+    torch::jit::Node* merged = mergeNodeIntoGroup(group, producer->node());
+    // remaining uses of this producer can occur because we allow
+    // fusion in cases where uses remain after the consumer
+    // if these exist, re-route them to the version of producer
+    // created in FusionGroup
+    size_t i = -1;
+    for (auto output : producer->node()->outputs()) {
+      ++i;
+      if (output->uses().size() == 0) {
+        continue;
+      }
+      getSubgraph(group).registerOutput(merged->outputs()[i]);
+      torch::jit::Value* new_output = group->addOutput();
+      new_output->copyMetadata(new_output);
+      aliasDb_->replaceWithNewValue(output, new_output);
+      output->replaceAllUsesWith(new_output);
+    }
+    producer->node()->destroy();
+    return group;
+  }
+
+  torch::jit::value_list sortReverseTopological(torch::jit::ArrayRef<torch::jit::Value*> inputs) {
```

Comment:
This function requires another round of refinement. It's mainly borrowed from Pytorch, if I change the naming too much here, it will be hard to map it back to Pytorch code when doing the refinement. Maybe let's keep it for now. :)

Reply:
sg! Maybe a todo comment somewhere?

Reply:
I will create item for tracking.